proposals/NNNN-swiftpm-additional-plugin-apis.md

```diff
@@ -0,0 +1,450 @@
+    /// Whether the library is static, dynamic, or automatically determined.
+    public let type: LibraryType
+}
+
+/// A type of library product.
+public enum LibraryType {
+    /// Static library.
+    case `static`
+
+    /// Dynamic library.
+    case `dynamic`
+
+    /// The type of library is unspecified and will be determined at build time.
+    case `automatic`
+}
+
+/// Represents a single target defined in a package. Specializations represent
+/// different types of targets.
+public class Target {
+    /// The name of the target, as defined in the package manifest. It's unique
+    /// among the targets of the package.
+    public var name: String
+    
+    /// The absolute path of the target directory in the local file system.
+    public var directory: Path
+    
+    /// Any other targets on which this target depends, in the same order as
+    /// they are specified in the package manifest. Conditional dependencies
+    /// that do not apply have already been filtered out.
+    public var dependencies: [Dependency]
+
+    /// Represents a dependency of a target on a product or on another target.
+    public enum Dependency {
+        /// A dependency on a target in the same package.
+        case target(Target)
+
+        /// A dependency on a product in another package.
+        case product(Product)
+    }
+}
+
+/// Represents a target consisting of a source code module, containing either
+/// Swift or source files in one of the C-based languages.
+public class SourceModuleTarget: Target {
+    /// The name of the module produced by the target (derived from the target
+    /// name, though future SwiftPM versions may allow this to be customized).
+    public var moduleName: String
+
+    /// The directory containing public C headers, if applicable. This will
+    /// only be set for targets that have Clang sources, and only if there is
+    /// a public headers directory.
+    public var publicHeadersDirectory: Path?
+
+    /// The source files that are associated with this target (any files that
+    /// have been excluded in the manifest have already been filtered out).
+    public var sourceFiles: FileList
+}
+
+/// Represents a target describing a library that is distributed as a binary.
+public class BinaryLibraryTarget: Target {
+    /// The library in the local file system.
+    public var libraryPath: Path
+}
+
+/// Represents a target describing a system library that is expected to be
+/// present on the host system.
+public class SystemLibraryTarget: Target {
+    /// The directory containing public C headers, if applicable.
+    public let publicHeadersDirectory: Path?
+  
+    /// The library in the local file system.
+    public let libraryPath: Path
+}
+
+/// Provides information about a list of files. The order is not defined
+/// but is guaranteed to be stable. This allows the implementation to be
+/// more efficient than a static file list.
+public struct FileList: Sequence {
```

Comment:
Makes sense.  This hasn’t actually changed at all since SE-0303 but I included this here for completeness, but I can see making it a protocol even so.